backend/wine_map/resources.py

```diff
@@ -24,15 +24,46 @@ class Meta:
             "percent_of_alcohol",
             "region",
         )
+        use_bulk = True
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.processed = set()
+        self.country_cache = {}
+        self.brand_cache = {}
 
     def before_import_row(self, row: OrderedDict, **kwargs: dict) -> None:
-        row["country"] = (
-            Country.objects.get_or_create(name=row["country"])[0].id
-            if row["country"]
-            else None
-        )
-        row["brand"] = (
-            Brand.objects.get_or_create(name=row["brand"])[0].id
-            if row["country"]
-            else None
-        )
+        self.ensure_country_created(row)
+        self.ensure_brand_created(row)
+
+    def after_import_row(self, row: OrderedDict, row_result: RowResult,
+                         row_number: int = None, **kwargs) -> None:
+        self.processed.add((row["name"], row["brand"]))
+
+    def skip_row(self, instance: Wine, original: Wine, row: OrderedDict,
+                 import_validation_errors: bool = None) -> bool:
+        if not instance.name.strip():
+            return True
+        if (instance.name, instance.brand_id) in self.processed:
+            return True
+        return super().skip_row(instance, original, row, import_validation_errors)
```

Comment:
maybe make something like this:

```python
if instance.name.strip() and (instance.name, instance.brand_id) not in self.processed:
      return super().skip_row(instance, original, row, import_validation_errors)
return True
```

Reply:
I guess somethig like this will be easier to read
```suggestion
                 import_validation_errors: bool = None) -> bool:
        if not instance.name.strip() and (instance.name, instance.brand_id) in self.processed:
            return True
        return super().skip_row(instance, original, row, import_validation_errors)
```